Dapp explorer

dievazqu opened the pull request:
### Description

DApps Screen.
Designs: https://www.figma.com/file/zdK8XvxNXvTjxnM5V33pHw/dAPP-Directory?node-id=831%3A19659

### Tested

Manually in the Android and iOS emulator

### How others should test

Turn on the feature flag `dappsExplorerEnabled`, then you should be able to interact with the new screen.

### Known Issue
In the new Dapps screen, if the user presses on Bidali (which it's not visible yet) and then goes back, it doesn't return to the Dapps screen.

### Related issues

- Fixes #1506

Screenshots (this is a older version, some wording has changed):
<img width="493" alt="Screen Shot 2022-01-14 at 13 30 09" src="https://user-images.githubusercontent.com/12504748/149550602-a3f16d3f-be3c-4fc7-bc16-00fe30578dac.png">
<img width="502" alt="Screen Shot 2022-01-14 at 13 30 35" src="https://user-images.githubusercontent.com/12504748/149550680-0d944c4d-d4f6-41b1-891c-cd69861466a6.png">
<img width="497" alt="Screen Shot 2022-01-14 at 13 31 58" src="https://user-images.githubusercontent.com/12504748/149550846-17745e75-4996-4b90-95de-0c7dbc73e0ad.png">
<img width="496" alt="Screen Shot 2022-01-14 at 13 32 16" src="https://user-images.githubusercontent.com/12504748/149550885-a65eecd1-fada-4225-9bdb-bd40f4d0826d.png">

dievazqu commented on packages/mobile/src/components/BottomSheet.tsx:
```diff
@@ -0,0 +1,141 @@
+
+const MIN_EMPTY_SPACE = 100
+
+function BottomSheet({ children, isVisible, onBackgroundPress }: Props) {
```
Mainly I copy pasted this class from `TokenBottomSheet`.
I will create a next PR with a refactor to start using only this class instead of having both.

dievazqu commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,342 @@
+              {t('dappsScreenBottomSheet.message')}
+            </Text>
+
+            {/* <Text>Dont show again Message</Text> */}
```
There is a discussion regarding if we want this feature or not: https://valora-app.slack.com/archives/C02TH9GNERL/p1642102112018200

gnardini replied:
Lets delete this commented out stuff in the meantime, if we decide to add it we'll need to add extra logic anyways :)

gnardini commented on packages/mobile/locales/en-US/translation.json:
```diff
@@ -496,6 +496,7 @@
   "helpFindAccount": "Help your contact find their Account Address",
   "close": "Close",
   "home": "Home",
+  "dappsExplorer": "dAPP Explorer",
```
I think this text should be in the `base` directory as well?

dievazqu replied:
My bad, I think we can remove this one :)

gnardini commented on packages/mobile/src/app/reducers.ts:
```diff
@@ -85,6 +86,7 @@ const initialState = {
   multiTokenUseUpdatedFeed: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenUseUpdatedFeed,
   linkBankAccountEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.linkBankAccountEnabled,
   sentryTracesSampleRate: REMOTE_CONFIG_VALUES_DEFAULTS.sentryTracesSampleRate,
+  dappsExplorerEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.dappsExplorerEnabled,
```
nit: Can we put this variable next to `dappListApiUri `? Since both are related to the same thing :)

gnardini replied:
I think we also need to add `dappsExplorerEnabled` to the reducer for `Actions.UPDATE_REMOTE_CONFIG_VALUES` as well

gnardini commented on packages/mobile/src/icons/DappsExplorerLogo.tsx:
```diff
@@ -0,0 +1,78 @@
+}
+
+/*
+<Svg width="94" height="106" viewBox="0 0 94 106" fill="none" xmlns="http://www.w3.org/2000/svg">
```
What is this commented stuff? Can we delete it?

gnardini commented on packages/mobile/src/navigator/DrawerNavigator.tsx:
```diff
@@ -179,6 +181,8 @@ function CustomDrawerContent(props: DrawerContentComponentProps<DrawerContentOpt
 export default function DrawerNavigator() {
   const { t } = useTranslation()
   const isCeloEducationComplete = useSelector((state) => state.goldToken.educationCompleted)
+  const isDappsExplorerEnabled = useSelector((state) => state.app.dappsExplorerEnabled)
```
To avoid having to handle the error in the other screen, can we check as well if the URI is available? And if it's not not show the item since we know it will fail :)

dievazqu replied:
Yep, I'll do that

gnardini commented on packages/mobile/test/schemas.ts:
```diff
@@ -866,6 +866,7 @@ export const v26Schema = {
   app: {
     ...v25Schema.app,
     rampCashInButtonExpEnabled: false,
+    dappsExplorerEnabled: false,
```
I think that when we add new stuff we need to create a new schema, ie`v27Schema` with the new feature flag.

I think it will also require making a redux store migration.

gnardini commented on packages/mobile/src/components/BottomSheet.tsx:
```diff
@@ -0,0 +1,141 @@
+    borderTopRightRadius: Spacing.Regular16,
+    borderTopLeftRadius: Spacing.Regular16,
+  },
+  title: {
```
I like the new generic bottom sheet :) We should just remove the unused styles here

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+              dappUrl: app.url,
+            })
+          })
+          setcategoriesWithItems(Object.values(categoriesById))
```
typo: `setcategoriesWithItems` -> `setCategoriesWithItems`

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+  )
+}
+
+function renderItem(item: DappItem, onItemPress: (dapp: DappItem) => void) {
```
I think in general instead of items I'd call them `dApps` or something, in the categories object above as well, I think it's more descriptive

dievazqu replied:
It makes sense :)

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+          <Text style={styles.itemTitleText}>{item.name}</Text>
+          <Text style={styles.itemSubtitleText}>{item.description}</Text>
+        </View>
+        <Image source={require('src/images/link-arrow.png')} style={styles.linkArrow} />
```
I think for more icons we add the png in `src/images/Images.ts` and import it wherever it's used instead of requiring them inline like this.

Not sure why we do that, and I realize that there's one more place that does it like this, but I think it makes sense to copy the most common pattern?

dievazqu replied:
Yep, you are right, I just copy pasted but it makes sense to continue with the pattern

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+    )
+  }
+
+  const { loading } = useAsync(
```
One thing that's potentially missing is an error. If this fails there will be no items and we won't show the loading indicator anymore but no content either. We can ask Nitya but maybe a simple 'There was an error loading the dApps' could be enough? I'd check with her in any case.

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+              dappUrl: app.url,
+            })
+          })
+          setcategoriesWithItems(Object.values(categoriesById))
```
So this has the potential to cause an error (not sure if leak memory or what) because this is an async function that keeps a reference to the `setcategoriesWithItems` callback. If the user goes back or something this `setcategoriesWithItems` callback may not be defined anymore or not sure exactly what can happen. I think we can just return the `Object.values(categoriesById)` and it will be available in the `result` field that `useAsync` returns.

Maybe @jeanregisser can correct me if that's not the real problem or add more context 😅

dievazqu replied:
Oh, nice catch! I will change it, so it returns the value instead of calling `setCategoryiesWithItems`

jeanregisser replied:
Exactly @gnardini 👍

gnardini reviewed (commented):
Amazing work! Can't wait to see all this in action :)

kathaypacific commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+  const [dappSelected, setDappSelected] = useState<DappItem>()
+
+  if (!dappsListUrl) {
+    return (
```
i think this guy should be moved further down the component, just before the last `return` because if react will complain about mismatched number hooks between re-renders if the code does return early here :) (alternatively `useAsync` could be moved before this return, but it's easier to accidentally add another hook after this return later on by accident) (https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level)

dievazqu replied:
You are right, already changed :)

kathaypacific commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,349 @@
+  )
+}
+
+function renderCategoryWithItems(
```
was wondering if we have any preferences for patterns here...this way of course is fine and works :) but it is equivalent to doing `function CategoryWithItems(props) {...}` and calling it in line 198 like 
```
categoriesWithItems.map(
  <CategoryWithItems
    {...someProps}
  />
)}
```
this function is producing a react component so i personally prefer invoking it like a component but not sure what the consensus is, so thought i'd ask! we do use the `renderXXX` function pattern if the `renderXXX` is inside a component/class already which makes sense, but this one is outside and standalone.

dievazqu replied:
Yes! When I was doing it I thought that using react component would be better, I'll change it :)

kathaypacific reviewed (commented):
Nice work!!!! 🔥

jeanregisser commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,362 @@
+          {!loading &&
+            !error &&
+            result &&
+            result.map((category: CategoryWithDapps) => (
+              <CategoryWithDapps category={category} onDappPress={onDappPress} />
+            ))}
```
This is ok when we have few Dapps / Categories, but will become a performance issue if we have more than 10 or so of them. At that point we should favor a `FlatList`.

dievazqu replied:
It makes sense, I will change it

jeanregisser replied:
Or ˋSectionList`

jeanregisser commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,362 @@
+    alignItems: 'flex-start',
+    justifyContent: 'center',
+    flex: 1,
+    marginVertical: 16,
```
We haven't used this everywhere sofar, but we have standard spacing constants. For instance: https://github.com/valora-inc/wallet/blob/c1e0fae65108e064ce313f56ae85cbba84b1810f/packages/mobile/src/verify/VerificationLoadingScreen.tsx#L290

jeanregisser commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,362 @@
+  itemTitleText: {
+    ...fontStyles.small,
+    fontWeight: '500',
+    color: '#2E3338',
```
Can't we use the colors from our design system?

https://github.com/valora-inc/wallet/blob/6bf13e1bb5a3cfda63eebf0477f829430085e598/packages/react-components/styles/colors.tsx#L1-L36

dievazqu replied:
Yes, we can! Thanks!

jeanregisser commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,362 @@
+    paddingVertical: 4,
+  },
+  itemTitleText: {
+    ...fontStyles.small,
+    fontWeight: '500',
```
I forgot the exact details now, but there are differences between iOS and Android related to `fontWeight` and our custom font, so this might not render the way we expect.
We should stick to the styles defined in our design system to avoid this.

https://github.com/valora-inc/wallet/blob/66cdc5a6bd34cd0957e968d599e223cccaf6496e/packages/react-components/styles/fonts.tsx#L38-L115

dievazqu replied:
I added that to honor the design parameters. But I just removed the `fontWeight` property and it's showing the same, so we can remove it.

jeanregisser commented on packages/mobile/src/app/selectors.ts:
```diff
@@ -107,3 +107,5 @@ export const multiTokenUseUpdatedFeedSelector = (state: RootState) =>
 
 export const sentryTracesSampleRateSelector = (state: RootState) => state.app.sentryTracesSampleRate
+
+export const dappsListApiUrlSelector = (state: RootState) => state.app.dappListApiUri
```
nit: So should we use `uri` or `url`? 😄 
Looks like we've already mixed both, seems `url` should be preferred in my opinion.

jeanregisser reviewed (commented):
Looking great!! 🤩 
Left mostly some small nits/comments.

dievazqu commented on packages/mobile/src/components/Dialog.tsx:
```diff
@@ -17,8 +17,8 @@ interface Props {
   children: React.ReactNode
-  actionText: string
-  actionPress: () => void
+  actionText?: string
+  actionPress?: () => void
```
Well, I wasn't sure about this.
The primary action has some style that is different than the secondary action (green color text instead of gray).
In this case what we want is a Dialog that can be dismissed, and I think this will be a normal use case. That's why I allowed this behavior, but if you don't agree, what we can do is pass some style for the actionText.
However, I think it's great to have the possibility to use the Dialog in this case. Then it would be developer's responsibility to see how it's the best way to use the Dialog.

jeanregisser replied:
ah I see, then maybe we could have add a `highlightPrimaryAction?: boolean`, which would default to `true`.
This way the highlighted style on the primary action would be configurable and we would ensure there's an action handler passed.
What do you think?

dievazqu replied:
Hmm... I'm not sure. What will be the expected behavior if `highlightPrimaryAction = false` and we also pass a secondaryAction?
I think it's ok to have the possibility to create a Dialog without any action. Why should we force it to have one?

I'm gonna try to merge this, but we can continue this discussion offline if you want (or we can keep using this thread), and I will create a follow up PR if needed, that sounds ok?

jeanregisser replied:
Yes that's fine.

dievazqu commented on packages/mobile/src/icons/LinkArrow.tsx:
```diff
@@ -1,8 +1,8 @@
 
-const LinkArrow = () => {
-  return <Image source={require('src/images/link-arrow.png')} style={{ width: 32, height: 32 }} />
+const LinkArrow = ({ style = { width: 32, height: 32 } }) => {
+  return <Image source={require('src/images/link-arrow.png')} style={style} />
```
I like this, thanks!

dievazqu commented:
@gnardini @kathaypacific @jeanregisser 
Thank you for your reviews!
I think I have addressed all the comments, and the PR is ready to be re-review again. 😃

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,389 @@
+
+        return Object.values(categoriesById)
+      } catch (error) {
+        Logger.error(TAG, 'onError', error as Error)
```
Maybe add some small detail to the message? Like `Error fetching dApp list` instead of `onError`

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,389 @@
+            </View>
+          )}
+          {!loading && !error && result && (
+            <SectionList
```
I'm not sure about having the `SectionList` inside the `ScrollView` since they are both scrollable, I think the containing `ScrollView` should be just a View? I'm not sure how the interaction works though, mentioning it just in case.

dievazqu replied:
I'll try removing the ScrollView and see how that works :)

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,389 @@
+}
+
+function CategoryHeader({ category }: { category: CategoryWithDapps }) {
+  Logger.debug(TAG, `Render category ${JSON.stringify(category)}`)
```
Is this log helpful? It looks like it would be a bit spammy since it's logged on every re-render, so it would keep logging the same thing over and over.

dievazqu replied:
I found it helpful while developing. But maybe we can delete them now.
It's true that it's logged on every re-render, however, there shouldn't be any of them, except the first one (after fetching the json file). So I also don't find it spammy.
On my opinion I will let them there. But happy to delete them if you still think we should.

dievazqu replied:
I removed them 😋

gnardini replied:
Oh, sorry about that, it's true that it's not re-rendered that often, so I would have been fine if you preferred to keep them, but I guess it's fine either way :)

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,389 @@
+}
+
+function Dapp({ dapp, onPressDapp }: DappProps) {
+  Logger.debug(TAG, `Render item ${JSON.stringify(dapp)}`)
```
Same related to the frequency of logging

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,389 @@
+function Dapp({ dapp, onPressDapp }: DappProps) {
+  Logger.debug(TAG, `Render item ${JSON.stringify(dapp)}`)
+
+  const onPress = () => {
```
super nit: I think this would look slightly nicer as `const onPress = () => onPressDapp(dapp)`

gnardini reviewed (commented):
Left a couple final nits, looks great :)

dievazqu commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,381 @@
+  )
+
+  const onPressDapp = (dapp: Dapp) => {
+    setDappSelected(dapp)
```
Hmm, you are right. Let me see what I can do. Nice catch!

kathaypacific commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,386 @@
+  bottomSheetHeader: {
+    flex: 1,
+    flexDirection: 'row-reverse',
+    alignItems: 'flex-start',
```
super nit: i think you can have `alignItems: 'flex-end'` and remove the flexDirection to achieve the same effect?

kathaypacific replied:
actually you could even just add `style={{ alignSelf: 'flex-end' }}` to the `TopBarIconButton` and remove the need for the surrounding `View` too

dievazqu replied:
I think you are right, let me try it. Thanks!

kathaypacific commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,386 @@
+          backgroundColor: category.backgroundColor,
+        }}
+      >
+        <Text style={{ ...styles.categoryText, color: category.fontColor }}>{category.name}</Text>
```
i notice that there are a lot of inline styles in this component, do we want to keep them like this or move to a stylesheet? i think having styles in the stylesheet has some performance benefits (not recreating the style every rerender) and slightly easier to read but i see there are other places in the app where we use inline style so i'm not sure! just thought i'd ask, what do you think?

jeanregisser replied:
We're trying to avoid inlining whenever possible. But I guess here it would look like this, which is really equivalent:
```suggestion
        <Text style={[styles.categoryText, { color: category.fontColor }]}>{category.name}</Text>
```

kathaypacific commented on packages/mobile/src/redux/migrations.ts:
```diff
@@ -362,4 +362,11 @@ export const migrations = {
+    ...state,
+    app: {
+      ...state.app,
+      dappsExplorerEnabled: false,
```
i was expecting to see `dappListApiUrl` here, maybe nice to add it for correctness? (i also notice `dappListApiUrl` was added to the v7 schema in the test schemas 🙈)

dievazqu replied:
Oh, it makes sense, I'll fix it.

kathaypacific reviewed (approved):
This is awesome Diego!! Looks like you've been doing RN for years 🔥 I just left a few super small comments but honestly, looks great to me 🚀

gnardini commented on packages/mobile/src/dappsExplorer/DAppsExplorerScreen.tsx:
```diff
@@ -0,0 +1,386 @@
+  const onPressNavigationButton = () => {
+    if (!dappSelected) {
+      // Should never happen
+      Logger.warn(TAG, 'Internal error. There was no dapp selected')
```
Should this be `Logger.error` instead? I'm sorry, just now saw it :(

dievazqu replied:
Sure

gnardini reviewed (approved):
I think the conflicts are my fault, sorry about that :( Looks amazing, can't wait to have this in-app!

ValoraQA commented:
Hey @MuckT we have verified the above task on Latest iOS Test Flight Release Build V:1.25.1 and updated the status in test rails

**Test rail link** :
**iOS 13:** https://valoraapp.testrail.io/index.php?/tests/view/4401&group_by=cases:section_id&group_order=asc&group_id=30

[Bug link
](https://github.com/valora-inc/wallet/issues/1818)
**Devices** : iPhone 7 (13.3)
cc : @dievazqu 
Thanks..!

ValoraQA commented:
Hey @MuckT, we have verified the above task on Latest iOS Test Flight Alfajores Build V:1.25.1 & updated the status in test rails

**Test rail link :**
**iOS 15:**https://valoraapp.testrail.io/index.php?/tests/view/4464

**Observation :** We are able to view Dapp section in Hamburger menu and we are also able to access the page.

**Bug:** [Link](https://github.com/valora-inc/wallet/issues/1818)
**Devices :** iPhone 13 mini (15.1.1)
Cc : @dievazqu
Thanks..!!

ValoraQA commented:
Hey @MuckT, we have verified the above task on Latest Android Alfajores Build V:1.25.2 & updated the status in test rails

**Test rail link:**
**Android 10:** https://valoraapp.testrail.io/index.php?/tests/view/7082&group_by=cases:section_id&group_order=asc&group_id=30

**Observation**: We are able to view Dapp section in Hamburger menu and we are also able to access the page.

[Bug link
](https://github.com/valora-inc/wallet/issues/1818)

**Devices** : Samsung Galaxy M40(10.0)

**Cc:** @dievazqu 
Thanks..!!

ValoraQA commented:
Hey @MuckT, we have verified the above task on Latest Android Alfajores Build V:1.25.2 & updated the status in test rails


Android 8: 

https://valoraapp.testrail.io/index.php?/tests/view/7145&group_by=cases:section_id&group_order=asc&group_id=30

**Observation:** We are able to view Dapp section in Hamburger menu and we are also able to access the page.



Devices : Samsung Galaxy J5 Prime(8.0.0)

Cc: @dievazqu
Thanks..!!

ValoraQA commented:
Hey @MuckT, We have verified the above task on Latest Android Alfajores Build V:1.25.2 & updated the status in test rails. User is able to view Dapp section in Hamburger menu and we are also able to access the page.

**Test rail link:**
**Android 9:**  https://valoraapp.testrail.io/index.php?/tests/view/7469

**Bug:** [Link](https://github.com/valora-inc/wallet/issues/1818)
**Devices :** Redmi Note 5 Pro (9.0)
Cc: @dievazqu
Thanks..!!
